Do no delete default profiles if no rules were applied

AdrienClairembault opened the pull request:
When users have only one dynamic profiles that match the default profile of GLPI, this profile should not be deleted by the right rules engine if no rules match this user.

Concrete example (!21527) :  
- An user is created through the API
- The default profile is auto affected as dyamic
- There is no right rules targeting this user
-> When this user log in, he should keep the current profile (currently the profile is removed and he can't login).

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

trasher reviewed (commented):
Sounds good, tests should be added

trasher commented on tests/units/RuleRight.php:
```diff
@@ -152,4 +152,35 @@ public function testLocalAccount() {
+      ];
+
+      $user = new \User();
+      $users_id = $user->add($testuser);
```
```suggestion
      $users_id = $user->add($testuser);
      $this->integer($users_id)->isGreaterThan(0);
```

AdrienClairembault commented:
Rebased + forcepushed since my 9.5 branches was a bit too old.

As we discussed this morning, added a new "is_default_profile" field to handle case where the default profile was modified.

cedric-anne commented on inc/profile_user.class.php:
```diff
@@ -99,6 +99,13 @@ function prepareInputForAdd($input) {
          return false;
       }
+
+      if (isset($input['profiles_id']) && !empty($input['profiles_id'])) {
+         if ($input['profiles_id'] == Profile::getDefault()) {
+            $input['is_default_profile'] = true;
+         }
+      }
```
IMHO, this `is_default_profile` flag should only be added from the current `User::post_addItem()`. Indeed if this profile is added manually or via right rules, then this profile would have been explicitely added, and then should not be considered as a default value.

cedric-anne commented on inc/user.class.php:
```diff
@@ -1050,6 +1050,25 @@ function applyRightRules() {
             $return = true;
+         } else if (count($dynamic_profiles) == 1) {
+            $dynamic_profile = reset($dynamic_profiles);
+            $default_profile = Profile::getDefault();
```
This should be moved inside the `if ($dynamic_profile['is_default_profile'] == true) {` block, to prevent an useless query if this condition is not reached.

cedric-anne commented on tests/units/RuleRight.php:
```diff
@@ -152,4 +152,63 @@ public function testLocalAccount() {
+      $right2 = array_pop($pu);
+      $this->array($right2)->isEqualTo($right);
+
+      $_SESSION['glpiactiveprofile']['user'] = 7327;
+      $_SESSION['glpiactiveprofile']['profile'] = 23;
+
```
```suggestion
```

AdrienClairembault replied:
Does not work without, seems like there is some rights checks inside the DB engine for this case :/

cedric-anne replied:
Where does thes values comes ?

AdrienClairembault replied:
empty_data.php, profiles rights for a super admin:
![image](https://user-images.githubusercontent.com/42734840/108345705-e3980000-71de-11eb-9143-3045340e7265.png)

cedric-anne replied:
Use `$this->login();` to login with TU_USER which has these rights, instead of forcing value.

AdrienClairembault replied:
I've tried it before and it didn't work but it seemed weird since as you said TU_USER have the two rights that I was adding.
I just found out that you also need to be the one that created the user to be allowed to edit his profiles, so adding a second `$this->login();` before the creating the user at line 164 make it works.

Is it normal to have these rights issues ?
Shouldn't rights be managed in an upper layer (front files, controllers, ...) and not intervene in this kind of situations ?

cedric-anne replied:
> I just found out that you also need to be the one that created the user to be allowed to edit his profiles

Are you sure of that ?

I think it was failing before because, and it is normal, a user cannot add itself a profile that gives him more rights. As in your test, you were connected as the tested user `$this->login($testuser['name'], $testuser['password'], false);`, it is normal that you were not able to add a profile for this user.

cedric-anne commented on inc/user.class.php:
```diff
@@ -1050,6 +1050,25 @@ function applyRightRules() {
+               if ($dynamic_profile['profiles_id'] !== $default_profile) {
+                  $pu = new Profile_User();
+                  $dynamic_profile['profiles_id'] = $default_profile;
+                  $pu->add($dynamic_profile);
+                  $pu->delete([
+                     'id' => $dynamic_profile['id']
+                  ]);
```
Why did you not use an update instead ?

AdrienClairembault replied:
Update doesn't work due to some hidden right checks (same is done inside `applyRightRules`,  it doesn't update anything, only remove and add).
